WalletWasabi/Wallets/WalletManager.cs

```diff
@@ -85,47 +100,118 @@ public bool AnyWallet()
 			}
 		}
 
-		public async Task<Wallet> CreateAndStartWalletAsync(KeyManager keyManager)
+		public async Task<Wallet> StartWalletAsync(Wallet wallet)
 		{
-			using (await AddRemoveLock.LockAsync(CancelAllInitialization.Token).ConfigureAwait(false))
+			Guard.NotNull(nameof(wallet), wallet);
+
+			if (wallet.State == WalletState.Starting)
+			{
+				return wallet;
+			}
+
+			using (await StartStopWalletLock.LockAsync(CancelAllInitialization.Token).ConfigureAwait(false))
 			{
-				Wallet wallet = null;
 				try
 				{
-					wallet = new Wallet(Network, BitcoinStore, keyManager, Synchronizer, Nodes, WalletDirectories.WorkDir, ServiceConfiguration, FeeProvider, BitcoinCoreNode);
+					wallet.RegisterServices(BitcoinStore, Synchronizer, Nodes, ServiceConfiguration, FeeProvider, BitcoinCoreNode);
 
 					var cancel = CancelAllInitialization.Token;
-					lock (Lock)
-					{
-						Wallets.Add(wallet, new HashSet<uint256>());
-					}
-
 					Logger.LogInfo($"Starting {nameof(Wallet)}...");
 					await wallet.StartAsync(cancel).ConfigureAwait(false);
 					Logger.LogInfo($"{nameof(Wallet)} started.");
-
 					cancel.ThrowIfCancellationRequested();
 
-					wallet.WalletRelevantTransactionProcessed += TransactionProcessor_WalletRelevantTransactionProcessed;
-					wallet.OnDequeue += ChaumianClient_OnDequeue;
-
 					return wallet;
 				}
 				catch
 				{
-					if (wallet is { })
+					await wallet.StopAsync(CancellationToken.None).ConfigureAwait(false);
+					throw;
+				}
+			}
+		}
+
+		public Task<Wallet> StartWalletAsync(KeyManager keyManagerToFindByReference)
+		{
+			Wallet wallet;
+			lock (Lock)
+			{
+				wallet = Wallets.Single(x => x.Key.KeyManager == keyManagerToFindByReference).Key;
+			}
+
+			return StartWalletAsync(wallet);
+		}
+
+		public Task<Wallet> AddAndStartWalletAsync(KeyManager keyManager)
+		{
+			var wallet = AddWallet(keyManager);
+			return StartWalletAsync(wallet);
+		}
+
+		public Wallet AddWallet(KeyManager keyManager)
+		{
+			lock (Lock)
+			{
+				if (Wallets.Any(w => w.Key.KeyManager == keyManager))
+				{
+					throw new InvalidOperationException($"Wallet already added: {keyManager.FilePath}.");
+				}
+			}
+
+			Wallet wallet = new Wallet(WalletDirectories.WorkDir, Network, keyManager);
+			AddWallet(wallet);
+			return wallet;
+		}
+
+		public Wallet AddWallet(string walletName)
```

Comment:
This method is broken. In previous versions you were not able to open nonexistent wallets but in this version you can. Do this: 
1. Open your Wasabi 
2. In the Load wallet page choose a wallet and delete it from the Wallets and WalletBackups folder.
3. Open it.
**Expected**: A message alerting the wallet is not there.
**Actual**: It creates a new wallet.

Reply:
What do you mean? It throws: 
https://github.com/molnard/WalletWasabi/blob/c74ed79fa08e2e5de0d96e9f23f5c5ab5fd3ef26/WalletWasabi/Wallets/WalletManager.cs#L178

Reply:
I mean that it doesn't work. You can verify it by yourself because it is 100% reproducible. Check how it works in any previous versions (v1.1.10.2 is a good candidate) and how it doesn't work on this PR. 

When I test this PR I cannot get the expected error:

![image](https://user-images.githubusercontent.com/127973/77253848-f22b0f80-6c3b-11ea-8902-13e2ff05f681.png)

I saw this first in a PR from @nopara73 but I didn't test it at the time. When I saw it here I decided to verify it and yep, this is breaking an existent feature.

Reply:
Is there a valid case that this could happen without hacking? It works well IMO, you can still load the wallet furthermore on the exit it will write it back. 
I think it fine like this.

Reply:
Something is a feature or a bug depending on what the expected behavior is. We could say that users should not be able to open wallets that do not exist. We could also say that in case a wallet does not exist anymore then it should be recreated. Both are okay.

Honestly, in this case I don't remember why we added those checks, it seems it was important to let the user know that the wallet listed is not there anymore, but I don't remember the specific reason. What I know is that the code was introduced, reviewed and tested and should not stop working without a previous agreement. 

What worries me is that this stopped working by accident and not because that was what we wanted to achieve.

Reply:
@nopara73 do you remember why that was required? I was thinking to add some `File.Exist` to some places to solve this but it does not seem to be fit our current logic. 
Do we need this?